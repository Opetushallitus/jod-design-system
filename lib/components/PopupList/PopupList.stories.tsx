import type { Meta, StoryObj } from '@storybook/react';

import { PopupList } from './PopupList';
import { fn } from '@storybook/test';

const meta = {
  title: 'PopupList',
  component: PopupList,
  tags: ['autodocs'],
} satisfies Meta<typeof PopupList>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  parameters: {
    docs: {
      description: {
        story: 'Default popup list behavior with links',
      },
    },
  },
  args: {
    items: [
      { label: 'Omat sivut', href: '#' },
      { label: 'Kirjaudu ulos', href: '#' },
    ],
  },
};

export const Active: Story = {
  parameters: {
    docs: {
      description: {
        story: 'Popup list with active indicator',
      },
    },
  },
  args: {
    items: [{ label: 'Suomeksi' }, { label: 'På svenska' }, { label: 'In English', active: true }],
  },
};

export const ButtonsAndLinks: Story = {
  parameters: {
    docs: {
      description: {
        story: 'Popup list with links and buttons',
      },
    },
  },
  args: {
    items: [
      { label: 'Luo tili', type: 'link', href: '#' },
      { label: 'Kirjaudu sisään', type: 'button', onClick: fn() },
    ],
  },
};
