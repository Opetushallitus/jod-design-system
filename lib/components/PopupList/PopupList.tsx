import { ActiveIndicator } from '../ActiveIndicator/ActiveIndicator';

export interface PopupListItem {
  /** List item label */
  label: string;
  /** Is list item active */
  active?: boolean;
  /** Optional click handler */
  onClick?: () => void;
  /** Link destination */
  href?: string;
  /** Type of the element. Default is link */
  type?: 'button' | 'link';
}

export interface PopupListProps {
  /** List items */
  items: PopupListItem[];
}

export const PopupList = ({ items }: PopupListProps) => {
  const hasActives = Array.isArray(items) && items.some((item) => item.active);
  return (
    <div className="inline-flex flex-col items-start rounded-[20px] border-[3px] border-[#767676] bg-white p-6 pr-10">
      <ul>
        {items.map((item) => {
          const Component = item.type === 'button' ? 'button' : 'a';

          return (
            <li key={item.label} className={hasActives ? 'pl-6' : ''}>
              <Component
                {...(Component === 'button' ? { type: 'button', onClick: item.onClick } : { href: item.href ?? '#' })}
                aria-current={item.active ? 'true' : undefined}
                className={`flex items-center gap-3 py-3 text-button-md hover:text-accent hover:underline focus:left-auto focus:underline`}
              >
                {item.active && <ActiveIndicator />}
                {item.label}
              </Component>
            </li>
          );
        })}
      </ul>
    </div>
  );
};
