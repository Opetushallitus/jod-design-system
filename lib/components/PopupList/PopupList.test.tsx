import { render, screen } from '@testing-library/react';
import { PopupList, PopupListItem, PopupListProps } from './PopupList';
import { describe, it, expect, vi } from 'vitest';

describe('PopupList', () => {
  const mockOnClick = vi.fn();

  const items: PopupListItem[] = [
    { label: 'Item 1', active: true, onClick: mockOnClick },
    { label: 'Item 2', onClick: mockOnClick },
    { label: 'Item 3', onClick: mockOnClick },
  ];

  it('renders the PopupList component correctly', () => {
    const props: PopupListProps = { items };
    const { container } = render(<PopupList {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the correct number of items', () => {
    const props: PopupListProps = { items };
    render(<PopupList {...props} />);
    const listItemElements = screen.getAllByRole('listitem');
    expect(listItemElements.length).toBe(items.length);
  });

  it('triggers onClick when an item is clicked', () => {
    const props: PopupListProps = { items };
    render(<PopupList {...props} />);
    const itemElement = screen.getByRole('button', { name: 'Item 2' });
    itemElement.click();
    expect(mockOnClick).toHaveBeenCalled();
  });

  it('renders items as buttons or links based on the itemsType prop', () => {
    const links: PopupListItem[] = [
      { label: 'Item 1', href: '/foo' },
      { label: 'Item 2', href: '/bar' },
    ];
    render(<PopupList items={links} itemsType="link" />);
    const linkItemElement = screen.getByRole('link', { name: 'Item 2' });
    expect(linkItemElement).toBeInTheDocument();
    expect(linkItemElement).toHaveAttribute('href', '/bar');
  });
});
