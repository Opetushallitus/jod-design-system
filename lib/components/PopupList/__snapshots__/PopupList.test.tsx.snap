// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopupList > renders the PopupList component correctly 1`] = `
<div>
  <div
    class="ds-inline-flex ds-flex-col ds-items-start ds-rounded ds-bg-white ds-py-6 ds-px-[20px] ds-shadow-border ds-w-[257px]"
  >
    <div
      class="ds-flex ds-items-center ds-gap-3 ds-py-3 ds-text-heading-4 hover:ds-text-accent hover:ds-underline focus:ds-left-auto focus:ds-underline ds-w-full ds-pl-5 ds-rounded"
    >
      <a
        href="/foo"
      >
        Item 1
      </a>
    </div>
    <div
      class="ds-flex ds-items-center ds-gap-3 ds-py-3 ds-text-heading-4 hover:ds-text-accent hover:ds-underline focus:ds-left-auto focus:ds-underline ds-w-full ds-pl-5 ds-rounded"
    >
      <a
        href="/bar"
      >
        Item 2
      </a>
    </div>
    <div
      class="ds-flex ds-items-center ds-gap-3 ds-py-3 ds-text-heading-4 hover:ds-text-accent hover:ds-underline focus:ds-left-auto focus:ds-underline ds-w-full ds-pl-5 ds-rounded"
    >
      <a
        href="/baz"
      >
        Item 3
      </a>
    </div>
  </div>
</div>
`;
